### PR TITLE
Fixed #16501 -- Added a unicode parameter to models.SlugField and forms....

### PR DESCRIPTION
**Pull request by kutenai**
This is a rebase and update of pull request #3729 

Added unicode support to URLify and prepopulate scripts.
Matched unicode regex with current one.

Fixed an indentation error

Renamed unicode parameter to allow_unicode

### CODE REVIEW

File: django/contrib/admin/options.py [hunk 1]
```diff
@@ -111,7 +111,7 @@ class BaseModelAdmin(six.with_metaclass(forms.MediaDefiningClass)):
 
     @classmethod
     def check(cls, model, **kwargs):
-        return cls.checks_class().check(cls, model, **kwargs)
+            return cls.checks_class().check(cls, model, **kwargs)
```
Comment (timgraham) [inline django/contrib/admin/options.py]: indent inadvertently modified

File: django/contrib/admin/options.py [hunk 2]
```diff
@@ -1744,6 +1745,9 @@ def __init__(self, parent_model, admin_site):
     def media(self):
         extra = '' if settings.DEBUG else '.min'
         js = ['jquery%s.js' % extra, 'jquery.init.js', 'inlines%s.js' % extra]
+        # TODO: Was removed in master, modified in patch.
```
Comment (timgraham) [inline django/contrib/admin/options.py]: We can remove this - it's always included above.

File: django/contrib/admin/static/admin/js/prepopulate.js [hunk 3]
```diff
@@ -1,11 +1,12 @@
 (function($) {
-    $.fn.prepopulate = function(dependencies, maxLength) {
+    $.fn.prepopulate = function(dependencies, maxLength, unicode) {
```
Comment (timgraham) [inline django/contrib/admin/static/admin/js/prepopulate.js]: let's call the new variable `allowUnicode`

File: django/db/models/fields/__init__.py [hunk 4]
```diff
@@ -2112,6 +2112,9 @@ def __init__(self, *args, **kwargs):
         # Set db_index=True unless it's been set manually.
         if 'db_index' not in kwargs:
             kwargs['db_index'] = True
+        self.unicode = kwargs.pop('unicode', False)
```
Comment (timgraham) [inline django/db/models/fields/__init__.py]: and let's call it `allow_unicode` on the Python side.

File: docs/ref/forms/fields.txt [hunk 5]
```diff
@@ -847,6 +847,17 @@ For each field, we describe the default widget used if you don't specify
...
    :class:`~django.db.models.SlugField` in forms.
 
+   Takes an optional parameter:
+
+   .. attribute:: unicode
+
+       A boolean instructing the field to accept unicode letters. Defaults to
+       ``False``.
+
+   .. versionchanged:: 1.8
```
Comment (timgraham) [inline docs/ref/forms/fields.txt]: 1.8 should be 1.9 now

File: docs/ref/validators.txt [hunk 6]
```diff
@@ -183,6 +183,15 @@ to, or in lieu of custom ``field.clean()`` methods.
     A :class:`RegexValidator` instance that ensures a value consists of only
     letters, numbers, underscores or hyphens.
 
+``validate_unicode_slug``
+-------------------------
+.. data:: validate_slug
```
Comment (timgraham) [inline docs/ref/validators.txt]: please add a blank line between ------ and .. data ...

File: docs/releases/1.8.txt [hunk 7]
```diff
@@ -358,6 +358,9 @@ Forms
 * You can now pass a callable that returns an iterable of choices when
   instantiating a :class:`~django.forms.ChoiceField`.
 
+* :class:`~django.forms.SlugField` now accepts a
```
Comment (timgraham) [inline docs/releases/1.8.txt]: please move to 1.9 release notes

File: tests/model_fields/tests.py [hunk 8]
```diff
@@ -847,14 +856,14 @@ def test_IntegerField(self):
...
     def test_IPAddressField(self):
-        lazy_func = lazy(lambda: '127.0.0.1', six.text_type)
-        self.assertIsInstance(
-            IPAddressField().get_prep_value(lazy_func()),
-            six.text_type)
-        lazy_func = lazy(lambda: 0, int)
-        self.assertIsInstance(
-            IPAddressField().get_prep_value(lazy_func()),
-            six.text_type)
+            lazy_func = lazy(lambda: '127.0.0.1', six.text_type)
```
Comment (timgraham) [inline tests/model_fields/tests.py]: inadvertent indent change

File: django/contrib/admin/options.py (hunk 2 above)
Comment (kutenai) [inline django/contrib/admin/options.py]: I do not see where this is added above. The previous line does not add it, so I'm wondering if there is something else missing.

Comment (timgraham) [inline django/contrib/admin/options.py]: The idea is that `InlineModelAdmin` is always used with a `ModelAdmin`, so we don't need to include the files again.

**Comment by kutenai** [comment]
I've obviously messed this up, and I'm working on resolving the issues. When I changed the attribute names, it caused a ripple effect that I"m working through.

**Comment by kutenai** [comment]
I believe I have addressed your issues now tim. I've also verified that the tests, docs, etc. run clean.

I had to add an unrelated fix to this file:
tests/gis_tests/distapp/tests.py

It was importing Point, but I don't have GIS, so for me it was failing. It was importing Point, which is not available if HAS_GIS is False.

This patch could use a final rebase when it is fully clean.

File: django/contrib/admin/options.py [hunk 9]
```diff
@@ -1744,8 +1745,6 @@ def __init__(self, parent_model, admin_site):
     def media(self):
         extra = '' if settings.DEBUG else '.min'
         js = ['jquery%s.js' % extra, 'jquery.init.js', 'inlines%s.js' % extra]
-        if self.filter_vertical or self.filter_horizontal:
```
Comment (timgraham) [inline django/contrib/admin/options.py]: don't think this should be removed as part of this patch?

File: django/contrib/admin/templates/admin/prepopulated_fields_js.html [hunk 10]
```diff
@@ -8,7 +8,8 @@
         id: '#{{ field.field.auto_id }}',
         dependency_ids: [],
         dependency_list: [],
-        maxLength: {{ field.field.field.max_length|default_if_none:"50"|unlocalize }}
+        maxLength: {{ field.field.field.max_length|default_if_none:"50"|unlocalize }},
+        unicode: {{ field.field.field.unicode|lower }}
```
Comment (timgraham) [inline django/contrib/admin/templates/admin/prepopulated_fields_js.html]: unicode -> allow_unicode, right? maybe we need to add a test for this...

File: django/contrib/staticfiles/handlers.py [hunk 11]
```diff
@@ -1,6 +1,7 @@
 from django.conf import settings
 from django.contrib.staticfiles import utils
 from django.contrib.staticfiles.views import serve
+from django.core import urlresolvers
```
Comment (timgraham) [inline django/contrib/staticfiles/handlers.py]: Think the changes to this file are inadvertent?

File: django/db/models/fields/__init__.py [hunk 12]
```diff
@@ -2122,13 +2125,15 @@ def deconstruct(self):
             kwargs['db_index'] = False
         else:
             del kwargs['db_index']
+        if self.allow_unicode is not False:
```
Comment (timgraham) [inline django/db/models/fields/__init__.py]: I think `if not self.allow_unicode` would be fine

File: tests/forms_tests/tests/test_fields.py [hunk 13]
```diff
@@ -1509,6 +1509,18 @@ def test_slugfield_normalization(self):
...
+    def test_slugfield_unicode_normalization(self):
+        f = SlugField(allow_unicode=True)
+        self.assertEqual(f.clean('a'), 'a')
+        self.assertEqual(f.clean('1'), '1')
+        self.assertEqual(f.clean('a1'), 'a1')
+        self.assertEqual(f.clean('你好'), '你好')
+        self.assertEqual(f.clean('  你-好  '), '你-好')
+        self.assertEqual(f.clean('ıçğüş'), 'ıçğüş')
+        self.assertEqual(f.clean('foo-ıç-bar'), 'foo-ıç-bar')
+        # Suggested in pull request, but questions remain
```
Comment (timgraham) [inline tests/forms_tests/tests/test_fields.py]: can you elaborate on what the issue is here?

File: tests/gis_tests/distapp/tests.py [hunk 14]
```diff
@@ -1,6 +1,6 @@
 from __future__ import unicode_literals
 
-from django.contrib.gis.geos import HAS_GEOS, Point
```
Comment (timgraham) [inline tests/gis_tests/distapp/tests.py]: Thanks, I committed this change separately so you can rebase your branch so it doesn't appear anymore.

File: django/contrib/staticfiles/handlers.py (hunk 11 above)
Comment (kutenai) [inline django/contrib/staticfiles/handlers.py]: I added it because urlresolvers was not imported, but it was used.

File: django/db/models/fields/__init__.py [hunk 15]
```diff
@@ -374,7 +374,7 @@ def deconstruct(self):
         Note that the positional or keyword arguments must contain values of the
         following types (including inner values of collection types):
 
-         * None, bool, str, unicode, int, long, float, complex, set, frozenset, list, tuple, dict
+         * None, bool, str, allow_unicode, int, long, float, complex, set, frozenset, list, tuple, dict
```
Comment (timgraham) [inline django/db/models/fields/__init__.py]: Revert change on this line (caused by find/replace, I guess).

File: django/contrib/staticfiles/handlers.py (hunk 11 above)
Comment (timgraham) [inline django/contrib/staticfiles/handlers.py]: I meant, what is the purpose of the changes? They don't appear to be part of the original patch. Are they related to supporting unicode in `SlugField`?

**Comment by timgraham** [comment]
@kutenai, do you have time/interest in completing this soon? I could offer it as a sprint task at DjangoconEU if not. Thanks!

**Comment by kutenai** [comment]
Actually, Yes.
I've been out of commission for a while due to family issues and work 
issues. A perfect storm if you will.

I'll get back on this very soon.. sorry for the delay

**Comment by kutenai** [comment]
All of the tests work for me locally, but fail here. I think I need a little help understanding the errors in the build.

**Comment by kutenai** [comment]
There is an error in master build. I just rebased against master so that is the source of the above error.

**Comment by timgraham** [comment]
Rebased & updated in https://github.com/django/django/pull/5010. Thanks!
